lib/Frontend/CompilerInvocation.cpp

```diff
@@ -973,6 +645,225 @@ StringRef FrontendArgsToOptionsConverter::determineFallbackModuleName() const {
+                                     bool useMainOutput) {
+    if (const Arg *A = Args.getLastArg(optWithPath)) {
+      Args.ClaimAllArgs(optWithoutPath);
+      output = A->getValue();
+      return;
+    }
+    
+    if (!Args.hasArg(optWithoutPath))
+      return;
+    
+    if (useMainOutput && !Opts.OutputFilenames.empty()) {
+      output = Opts.getSingleOutputFilename();
+      return;
+    }
+    
+    if (!output.empty())
+      return;
+    
+    llvm::SmallString<128> Path(Opts.originalPath());
+    llvm::sys::path::replace_extension(Path, extension);
+    output = Path.str();
+  };
+  
+  determineOutputFilename(Opts.DependenciesFilePath,
+                          OPT_emit_dependencies,
+                          OPT_emit_dependencies_path,
+                          "d", false);
+  determineOutputFilename(Opts.ReferenceDependenciesFilePath,
+                          OPT_emit_reference_dependencies,
+                          OPT_emit_reference_dependencies_path,
+                          "swiftdeps", false);
+  determineOutputFilename(Opts.SerializedDiagnosticsPath,
+                          OPT_serialize_diagnostics,
+                          OPT_serialize_diagnostics_path,
+                          "dia", false);
+  determineOutputFilename(Opts.ObjCHeaderOutputPath,
+                          OPT_emit_objc_header,
+                          OPT_emit_objc_header_path,
+                          "h", false);
+  determineOutputFilename(Opts.LoadedModuleTracePath,
+                          OPT_emit_loaded_module_trace,
+                          OPT_emit_loaded_module_trace_path,
+                          "trace.json", false);
+  
+  determineOutputFilename(Opts.TBDPath, OPT_emit_tbd, OPT_emit_tbd_path, "tbd",
+                          false);
+  
+  if (const Arg *A = Args.getLastArg(OPT_emit_fixits_path)) {
+    Opts.FixitsOutputPath = A->getValue();
+  }
+  
+  bool IsSIB = Opts.RequestedAction == FrontendOptions::EmitSIB ||
+  Opts.RequestedAction == FrontendOptions::EmitSIBGen;
+  bool canUseMainOutputForModule =
+  Opts.RequestedAction == FrontendOptions::MergeModules ||
+  Opts.RequestedAction == FrontendOptions::EmitModuleOnly || IsSIB;
+  auto ext = IsSIB ? SIB_EXTENSION : SERIALIZED_MODULE_EXTENSION;
+  auto sibOpt = Opts.RequestedAction == FrontendOptions::EmitSIB
+  ? OPT_emit_sib
+  : OPT_emit_sibgen;
+  determineOutputFilename(Opts.ModuleOutputPath,
+                          IsSIB ? sibOpt : OPT_emit_module,
+                          OPT_emit_module_path,
+                          ext,
+                          canUseMainOutputForModule);
+  
+  determineOutputFilename(Opts.ModuleDocOutputPath,
+                          OPT_emit_module_doc,
+                          OPT_emit_module_doc_path,
+                          SERIALIZED_MODULE_DOC_EXTENSION,
+                          false);
+}
+
+bool FrontendArgsToOptionsConverter::canEmitWhatActionCallsFor() const {
+  if (!Opts.canEmitDependencies()) {
```

Comment:
This doesn't really make sense. `canEmitDependencies` shouldn't check whether we actually *are* emitting dependencies.

Reply:
On second thought, I changed all of these to hasUnusable<something>Path. Let me know what you think.

Reply:
Argh! That's hasUnusable\<something\>Path.